### PR TITLE
Revamp dns-controller unit testing

### PR DESCRIPTION
**Pull request by shashidharatd**
This PR is reraised to k/f from k/k. The original PR is here https://github.com/kubernetes/kubernetes/pull/46606

This is just test improvement.

/cc @kubernetes/sig-multicluster-pr-reviews

### CODE REVIEW

**Comment by shashidharatd** [comment]
@irfanurrehman, this pr is rebased and all checks have passed. PTAL. Thanks!

File: pkg/dnsprovider/providers/google/clouddns/internal/stubs/managed_zone.go [hunk 1]
```diff
@@ -22,10 +22,9 @@ import "k8s.io/federation/pkg/dnsprovider/providers/google/clouddns/internal/int
 var _ interfaces.ManagedZone = ManagedZone{}
 
 type ManagedZone struct {
-	Service *ManagedZonesService
-	Name_   string
-	Id_     uint64
-	Rrsets  []ResourceRecordSet
+	Name_  string
```
Comment (irfanurrehman) [inline pkg/dnsprovider/providers/google/clouddns/internal/stubs/managed_zone.go]: [Clarification]: Am a little confused with the wordings of the commit and then the code; do u mean that u are making some of the methods re entrant?

File: pkg/dnsprovider/providers/google/clouddns/clouddns_test.go [hunk 2]
```diff
@@ -27,19 +27,19 @@ import (
...
+	return NewFakeInterface(zones) // Use this to stub out the entire cloud service
 }
 
 var interface_ dnsprovider.Interface
 
 func TestMain(m *testing.M) {
 	flag.Parse()
 	var err error
-	interface_, err = newTestInterface()
+	interface_, err = newTestInterface([]string{"example.com"})
```
Comment (irfanurrehman) [inline pkg/dnsprovider/providers/google/clouddns/clouddns_test.go]: Isn't the intention of making  NewFakeInterface() handle  multiple zones to test out passing multiple zones??

File: pkg/federation-controller/service/dns/dns_test.go [hunk 3]
```diff
@@ -290,17 +368,165 @@ func TestServiceController_ensureDnsRecords(t *testing.T) {
...
+			return false, nil
 		}
+		return true, nil
+	})
+	if err != nil {
+		t.Errorf("Actual DNS records does not match expected. Actual=%v, Expected=%v", availableDNSRecords, expectedDNSRecords)
 	}
 }
+
+func TestNewServiceDNSController(t *testing.T) {
```
Comment (irfanurrehman) [inline pkg/federation-controller/service/dns/dns_test.go]: [Clarification]: I did not clearly understand what additionally is this function testing, which is not already covered in TestServiceDNSController(). Is this mainly to improve code coverage?

Comment (shashidharatd) [reply pkg/federation-controller/service/dns/dns_test.go]: TestNewServiceDNSController is testing the initialisation parameters passed to the controller. The name is confusing, i will change it.

**Review by shashidharatd — COMMENTED** [review]
Thanks @irfanurrehman for review. i have addressed your comments. PTAL. Thanks !

**Pull request merged**